src/main/java/com/google/api/gax/retrying/AbstractRetryHandler.java

```diff
@@ -0,0 +1,137 @@
+ * @param <ResponseT> response type
+ */
+public abstract class AbstractRetryHandler<ResponseT> implements RetryHandler<ResponseT> {
+
+  private final NanoClock clock;
+
+  protected AbstractRetryHandler(NanoClock clock) {
+    this.clock = clock;
+  }
+
+  /**
+   * Ensures that the retry logic hasn't exceeded neither maximum number of retries nor the total
+   * execution timeout.
+   *
+   * @param e exception thrown by the previous attempt
+   * @param nextAttemptSettings attempt settings, which will be used for the next attempt, if
+   * accepted
+   * @return {@code true} if none of the retry limits are exceeded
+   */
+  @Override
+  public boolean accept(Throwable e, RetryAttemptSettings nextAttemptSettings) {
+    RetrySettings globalSettings = nextAttemptSettings.getGlobalSettings();
+    long randRetryDelayMillis = nextAttemptSettings.getRandomizedRetryDelay().getMillis();
+    long totalTimeSpentNanos =
+        clock.nanoTime()
+            - nextAttemptSettings.getFirstAttemptStartTime()
+            + TimeUnit.NANOSECONDS.convert(randRetryDelayMillis, TimeUnit.MILLISECONDS);
+
+    long totalTimeoutMillis = globalSettings.getTotalTimeout().getMillis();
+    long totalTimeoutNanos =
+        TimeUnit.NANOSECONDS.convert(totalTimeoutMillis, TimeUnit.MILLISECONDS);
+
+    return totalTimeSpentNanos <= totalTimeoutNanos
+        && (globalSettings.getMaxAttempts() <= 0
+            || nextAttemptSettings.getAttemptCount() < globalSettings.getMaxAttempts());
+  }
+
+  /**
+   * Crates next attempt settings. It increments the current attempt count and uses randomized
+   * exponential backoff factor for calculating next attempt execution time.
+   *
+   * @param e exception thrown by the previous attempt
+   * @param prevSettings previous attempt settings
+   * @return next attempt settings
+   */
+  @Override
+  public RetryAttemptSettings createNextAttemptSettings(
+      Throwable e, RetryAttemptSettings prevSettings) {
+    RetrySettings settings = prevSettings.getGlobalSettings();
+
+    long newRetryDelay = settings.getInitialRetryDelay().getMillis();
+    long newRpcTimeout = settings.getInitialRpcTimeout().getMillis();
+
+    if (prevSettings.getAttemptCount() > 0) {
+      newRetryDelay =
+          (long) (settings.getRetryDelayMultiplier() * prevSettings.getRetryDelay().getMillis());
+      newRetryDelay = Math.min(newRetryDelay, settings.getMaxRetryDelay().getMillis());
+      newRpcTimeout =
+          (long) (settings.getRpcTimeoutMultiplier() * prevSettings.getRpcTimeout().getMillis());
+      newRpcTimeout = Math.min(newRpcTimeout, settings.getMaxRpcTimeout().getMillis());
+    }
+
+    return new RetryAttemptSettings(
+        prevSettings.getGlobalSettings(),
+        Duration.millis(newRetryDelay),
+        Duration.millis(newRpcTimeout),
+        Duration.millis(ThreadLocalRandom.current().nextLong(newRetryDelay)),
+        prevSettings.getAttemptCount() + 1,
+        prevSettings.getFirstAttemptStartTime());
+  }
```

Comment:
Please check the previous comment. I really dislike putting logic into those type of classes. That class contains only configuraiton parameters of a concrete retry and nothing more, the name settings usually implies that, and is consistent with the rest of code (for example with RetrySettings).